File: algorithms/go/1743-restore-the-array-from-adjacent-pairs/restore-the-array-from-adjacent-pairs.go
package problem1743

// restoreArray 哈希表
func restoreArray(adjacentPairs [][]int) []int {
	mp := map[int][]int{}
	n := len(adjacentPairs) + 1
	for i := 0; i < n-1; i++ {
		pair := adjacentPairs[i]
		mp[pair[0]] = append(mp[pair[0]], pair[1])
		mp[pair[1]] = append(mp[pair[1]], pair[0])
	}

	ret := make([]int, n, n)

	// 找到首或者尾
	for key, arr := range mp {
		if len(arr) == 1 {
			ret[0] = key
			break
		}
	}

	ret[1] = mp[ret[0]][0]
	for i := 2; i < n; i++ {
		adj := mp[ret[i-1]]
		if ret[i-2] == adj[0] {
			ret[i] = adj[1]
		} else {
			ret[i] = adj[0]
		}
	}
	return ret
}

File: algorithms/1743-restore-the-array-from-adjacent-pairs/restore-the-array-from-adjacent-pairs.py
"""
1743. restore the array from adjacent pairs

https://leetcode.com/problems/restore-the-array-from-adjacent-pairs/
https://leetcode-cn.com/problems/restore-the-array-from-adjacent-pairs/

从相邻元素对还原数组
"""

from typing import List
import collections


def restore_array(adjacentPairs: List[List[int]]) -> List[int]:
    """
    哈希表
    """
    mp = collections.defaultdict(list)
    for pair in adjacentPairs:
        mp[pair[0]].append(pair[1])
        mp[pair[1]].append(pair[0])

    n = len(adjacentPairs) + 1
    ret = [0] * n

    # 找到首或尾
    for key, arr in mp.items():
        if len(arr) == 1:
            ret[0] = key
            break

    ret[1] = mp[ret[0]][0]
    for i in range(2, n):
        adj = mp[ret[i - 1]]
        ret[i] = adj[1] if adj[0] == ret[i - 2] else adj[0]

    return ret
